In [41]:
# 1 Base python
sec = 60 * 60 * 24 * 365
print(sec)

31536000


In [42]:
lifeTime = sec * 75
print(lifeTime)

2365200000


In [43]:
myAge = sec * 20
print(myAge)

630720000


In [44]:
print(myAge > 700000000)

False


In [45]:
Computing51 = 0.4 * (1 - 0.4)
print(Computing51)

0.24


In [46]:
Computing52 = (3 / 2) * (1 + 2.5 / 2) ** -4
print(Computing52)

0.058527663465935076


In [47]:
l1:list = ["The Shawshank Redemption", "The Godfather", "The Dark Knight", "The Godfather Part II", "12 Angry Men", "Schindler's List"]
print(l1)

['The Shawshank Redemption', 'The Godfather', 'The Dark Knight', 'The Godfather Part II', '12 Angry Men', "Schindler's List"]


In [48]:
l2 = l1[:3]
print(l2)

['The Shawshank Redemption', 'The Godfather', 'The Dark Knight']


In [49]:
l3 = l1[:2] + l1[4:]
print(l3)

['The Shawshank Redemption', 'The Godfather', '12 Angry Men', "Schindler's List"]


In [50]:
l4 = l1[::-1]
print(l4)

["Schindler's List", '12 Angry Men', 'The Godfather Part II', 'The Dark Knight', 'The Godfather', 'The Shawshank Redemption']


In [51]:
looplist:list = []
for i in range(70,80):
    looplist.append(i)
print(looplist)

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]


In [52]:
result = 0
count = 0
for i in looplist:
    result += i
    count += 1
print(result/count)

74.5


In [53]:
result = 0
for i in range(101):
    result += i
print(result)

5050


In [54]:
name = ["Adam", "Ashin", "Inukai", "Tanaka", "Ikki"]
number = ["33", "12", "45", "2", "17"]
for i in range(5):
    print(name[i] + ": " + number[i])

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


In [55]:
text = "Maps are data structures that contain key-value pairs. The python versions are called dicts. Such structures are often used to assign names to values in data, or to create complex data structures. The syntax is the following: {key1:value1, key2:value2, ...}. For instance, we can create a dict of squares of numbers:"
text = text.lower()
text = text.split()
wordCount = {}
for i in text:
    if i not in wordCount:
        wordCount[i] = 1
    else:
        wordCount[i] += 1
print(wordCount)

{'maps': 1, 'are': 3, 'data': 2, 'structures': 2, 'that': 1, 'contain': 1, 'key-value': 1, 'pairs.': 1, 'the': 3, 'python': 1, 'versions': 1, 'called': 1, 'dicts.': 1, 'such': 1, 'often': 1, 'used': 1, 'to': 3, 'assign': 1, 'names': 1, 'values': 1, 'in': 1, 'data,': 1, 'or': 1, 'create': 2, 'complex': 1, 'structures.': 1, 'syntax': 1, 'is': 1, 'following:': 1, '{key1:value1,': 1, 'key2:value2,': 1, '...}.': 1, 'for': 1, 'instance,': 1, 'we': 1, 'can': 1, 'a': 1, 'dict': 1, 'of': 2, 'squares': 1, 'numbers:': 1}


In [56]:
def timeCal(x:int) -> int:
    result = x // 100 + 0.01 * (x % 100) * 5 / 3
    return result

In [57]:
print(timeCal(1015))

10.25


In [58]:
# 2 Basic data frames

In [59]:
import pandas as pd

df = pd.read_csv('/Users/yanjiarui/desktop/INFO370/nycflights13.csv.bz2', sep='\t')

In [60]:
len(df.columns)

19

In [61]:
len(df)

336776

In [62]:
df.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
2      1089     5      40  2013-01-01T10:00:00Z  
3      1576     5      45  2013-01-01T10:00:00Z  
4       762     6       0  2013-01-01T11:00:00Z

It looks pretty reasonable. The time, carrier, flight, origin and dest, distance, are great.
The units of the dep time and arr time are not clear, by reading through the documentation, they are in HHMM or HMM form. The hour and minute seems to be wrong time.

In [63]:
df['origin'].nunique()

3

In [64]:
print(df['origin'].unique())

['EWR' 'LGA' 'JFK']


There are three airports in NYC: ['EWR' 'LGA' 'JFK']

In [65]:
len(pd.unique(df[df['year'] == 2013]['dest']))

105

There are 105 airports did the airlines fly form NYC in 2013

In [66]:
len(df[df['dest'] == 'SEA'])

3923

There are 3923 flights were from NYC to Seattle

In [67]:
len(df[df['dest'] == 'GEG'])

0

There aren't any flights from NYC to Spokane

In [68]:
df = df.copy()
df['duration_in_hours'] = (df['arr_time'] % 100 - df['dep_time'] % 100) / 60 + (df['arr_time'] // 100 - df['dep_time'] // 100)
df.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  duration_in_hours  
0      1400     5      15  2013-01-01T10:00:00Z           3.216667  
1      1416     5      29  2013-01-01T10:00:00Z           3.283333  
2      1089     5      40  2013-01-01T10:00:00Z           3.683333  
3      1576     5      45  2013-01-01T10:00:00Z           4.333333  
4       762     6       0  2013-01-01T11:00:00Z           2.300000

In [69]:
min(df['duration_in_hours'])

-23.35

In [70]:
df[df['duration_in_hours'] < 0].head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
719  2013      1    1    1929.0            1920        9.0       3.0   
725  2013      1    1    1939.0            1840       59.0      29.0   
791  2013      1    1    2058.0            2100       -2.0       8.0   
794  2013      1    1    2102.0            2108       -6.0     146.0   
797  2013      1    1    2108.0            2057       11.0      25.0   

     sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
719               7       -4.0      UA    1071  N27205    EWR  BQN     192.0   
725            2151        NaN      9E    3325  N905XJ    JFK  DFW       NaN   
791            2359        9.0      UA    1241  N27724    EWR  TPA     159.0   
794             158      -12.0      UA    1180  N78511    EWR  SJU     199.0   
797              39      -14.0      UA    1517  N17245    EWR  SFO     354.0   

     distance  hour  minute             time_hour  duration_in_hours  
719      1585    19      20  2013-01-02T00:00:00Z         -19.433333  
725      1391    18      40  2013-01-01T23:00:00Z         -19.166667  
791       997    21       0  2013-01-02T02:00:00Z         -20.833333  
794      1608    21       8  2013-01-02T02:00:00Z         -19.266667  
797      2565    20      57  2013-01-02T01:00:00Z         -20.716667

The minimum value of the duration I computed is -23.35.
The reason is because there are some flights depart at late night and arrive on tmr's morning. The computation is not able to calculate the flights last from the first day to the next day.

In [71]:
df_delay = df[['dep_delay','arr_delay']]

In [72]:
df_delay.sample(n = 10)

dep_delay  arr_delay
299448       -7.0      -27.0
312630      -10.0      -15.0
21346         3.0      -16.0
134464        8.0       51.0
58809        -2.0       -3.0
148588       -4.0       18.0
144110        4.0      -12.0
56360        -5.0      -36.0
10036        -3.0        2.0
322761       -7.0      -18.0

In [73]:
df_delay.dtypes

dep_delay    float64
arr_delay    float64
dtype: object

They are both float64 type

In [74]:
len(df_delay[(df_delay['dep_delay'] < 0) & (df_delay['arr_delay'] > 0)])

35648

There are 35648 flights leave early but arrive late

In [75]:
extra_delay = df_delay['arr_delay'] - df_delay['dep_delay']
extra_delay.max()

196.0

The longest “extra flight time” is 196.

In [76]:
len(df_delay[df_delay['arr_delay'] > 0])

133004

There are 133004 flights were delayed

In [77]:
len(df_delay[df_delay['arr_delay'] > 15])/len(df_delay)

0.2305092999501152

There are 23.05% flights were delayed by more than 15 minutes